### PR TITLE
feat: implement automatic table name detection for SQL export formats

### LINKED ISSUE #429: Support automatic table name detection for SQL export formats
## Summary
Implement automatic table name detection from SELECT queries for SQL export formats, eliminating the need to manually set CLI_SQL_TABLE_NAME for simple queries.

## Parent Issue
Part of #424 (Support SQL export formats for data migration)

## Relationship with DUMP TABLE (#425)
While DUMP TABLE (#425) will cover most typical use cases for full table exports, automatic table name detection provides additional flexibility:
- **Filtered exports**: `SELECT * FROM Users WHERE created_at > '2024-01-01'`
- **Sorted exports**: `SELECT * FROM Events ORDER BY timestamp DESC`
- **Limited exports**: `SELECT * FROM Logs LIMIT 1000` for testing
- **Custom conditions**: Complex WHERE clauses for partial data migration

DUMP TABLE is the recommended approach for full table dumps, but this feature enables SQL export for custom query results.

### Supported Query Patterns (Whitelist)
Only the following patterns will be supported for automatic detection:
- `SELECT * FROM table_name`
- `SELECT * FROM table_name WHERE ...`
- `SELECT * FROM table_name ORDER BY ...`
- `SELECT * FROM table_name LIMIT ...`
- Combinations of the above

Schema-qualified names are also supported:
- `SELECT * FROM schema.table_name WHERE ...`

### Explicitly NOT Supported
- SELECT with column list (risk of NOT NULL constraint violations)
- JOINs, subqueries, UNIONs, or other complex queries
- Table aliases
- CTEs (WITH clauses)
- Any query not matching the whitelist patterns

### Implementation Approach
1. Parse query using memefish AST
2. Match against whitelist patterns only
3. Extract table name as *ast.Path for proper identifier handling
4. If pattern doesn't match, require explicit CLI_SQL_TABLE_NAME setting

### Use Case Examples
```sql
-- Export only active users
SET CLI_FORMAT = 'SQL_INSERT_OR_UPDATE';
SELECT * FROM Users WHERE status = 'ACTIVE';

-- Export recent logs for debugging
SET CLI_FORMAT = 'SQL_INSERT';
SELECT * FROM AuditLogs WHERE timestamp > TIMESTAMP_SUB(CURRENT_TIMESTAMP(), INTERVAL 1 DAY);

-- Export sample data for testing
SET CLI_FORMAT = 'SQL_INSERT_OR_IGNORE';
SELECT * FROM Products ORDER BY RAND() LIMIT 100;
```

### Error Handling
- Clear error message when query doesn't match whitelist
- Suggest using DUMP TABLE for full table exports
- Suggest setting CLI_SQL_TABLE_NAME explicitly for complex queries

### Rationale for Conservative Approach
- SELECT * ensures all columns are included, avoiding NOT NULL violations
- Covers the gap between DUMP TABLE (full export) and manual CLI_SQL_TABLE_NAME (complex queries)
- Simpler implementation with fewer edge cases
- Users can always set CLI_SQL_TABLE_NAME explicitly for complex queries

### PR DESCRIPTION
**Pull request by apstndb**
## Summary

Implements automatic table name detection from SELECT queries for SQL export formats (INSERT, INSERT OR UPDATE, INSERT OR IGNORE), addressing #429.

This feature eliminates the need to manually set `CLI_SQL_TABLE_NAME` for simple SELECT queries while maintaining safety through strict pattern validation.

### Supported Query Patterns

The auto-detection supports only queries that preserve the original table structure:
- `SELECT * FROM table_name`
- `SELECT * FROM table_name WHERE ...`
- `SELECT * FROM table_name ORDER BY ...`
- `SELECT * FROM table_name LIMIT ...`
- `SELECT DISTINCT * FROM ...` (see insights below)
- Combinations of the above
- Query hints (e.g., `@{FORCE_INDEX=...}`)
- Table aliases (extracts actual table name)

### Explicitly Rejected Patterns

- **GROUP BY / HAVING** - Aggregations fundamentally change result set structure
- **JOINs** - Combine multiple tables
- **Specific column selection** - Risk of NOT NULL constraint violations in INSERT
- **Subqueries, CTEs, UNIONs** - Complex structures that don't map to single table

### Key Design Decisions & Insights

1. **DISTINCT Handling**: Initially rejected DISTINCT, but after discussion realized it's safe because:
   - All Spanner tables have primary keys (mandatory)
   - SELECT * results are inherently unique due to PKs
   - DISTINCT on unique results is effectively a no-op

2. **Scope Isolation**: Auto-detected table names are scoped to individual query executions:
   - Creates a copy of system variables for thread safety
   - Doesn't modify global `CLI_SQL_TABLE_NAME` setting
   - Each query's auto-detection is independent

3. **Error Handling**: Returns meaningful errors for debugging:
   - Logs reasons for auto-detection failure at DEBUG level
   - Helps users understand what patterns are supported
   - Falls back gracefully to requiring explicit `CLI_SQL_TABLE_NAME`

4. **AST Parsing with memefish**: Leverages the memefish SQL parser for reliable extraction:
   - Handles different AST structures (e.g., Query wrapper for ORDER BY/LIMIT)
   - Correctly processes schema-qualified names and reserved words
   - Validates query structure before extraction

### Testing

- Comprehensive unit tests covering all supported and rejected patterns
- Integration tests verifying end-to-end SQL export with auto-detection
- Tests for edge cases (hints, aliases, DISTINCT, etc.)

### User Experience

Before:
```sql
SET CLI_SQL_TABLE_NAME = 'Users';
SET CLI_FORMAT = 'sql_insert';
SELECT * FROM Users WHERE age > 18;
```

After:
```sql
SET CLI_FORMAT = 'sql_insert';
SELECT * FROM Users WHERE age > 18;
-- Table name automatically detected\!
```

Or using command-line flags:
```bash
# Before
spanner-mycli --format=sql_insert --sql-table-name=Users

# After (table name auto-detected from query)
spanner-mycli --format=sql_insert
```

For complex queries, users still have options:
1. Use `DUMP TABLE` for full table exports
2. Set `CLI_SQL_TABLE_NAME` explicitly for complex queries
3. Simplify query to match supported patterns

## Backward Compatibility

Fully backward compatible - explicit `CLI_SQL_TABLE_NAME` settings continue to work and take precedence over auto-detection.

Fixes #429

### CODE REVIEW

File: execute_sql.go [hunk 1]
```diff
@@ -98,6 +98,25 @@ func executeSQLImplWithVars(ctx context.Context, session *Session, sql string, s
...
+				// 3. Preserving user settings: the original CLI_SQL_TABLE_NAME remains unchanged
+				tempVars := *sysVars
+				tempVars.SQLTableName = detectedTableName
+				sysVars = &tempVars
+				slog.Debug("Auto-detected table name for SQL export", "table", detectedTableName)
+			} else if detectionErr != nil {
+				// Log why auto-detection failed for debugging
+				slog.Debug("Table name auto-detection failed", "reason", detectionErr.Error())
+			}
+		}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply execute_sql.go:119]: Fixed in a801765. Added SQLTableNameForExport field to Result struct to preserve the auto-detected table name through the entire execution pipeline, ensuring it works correctly in both streaming and buffered modes.

Fixed in commit a801765.

**Comment by apstndb** [comment]
## Potential Future Enhancements

### 1. Extended Pattern Support
While the current implementation is intentionally conservative, future versions could potentially support:
- `SELECT * FROM table WITH (NOLOCK)` style hints (if Spanner adds them)
- Multiple table detection for batch operations
- View name detection (once we understand view export semantics better)

### 2. Performance Optimization
The current implementation parses the query twice in SQL export mode:
1. Once by memefish for statement classification
2. Once by `extractTableNameFromQuery` for table detection

A future optimization could cache the parsed AST to avoid double parsing, though the performance impact is negligible for interactive use.

### 3. Column Mapping Support
A more sophisticated version could support column selection with mapping:
```sql
-- Future possibility: explicit column mapping
SET CLI_SQL_COLUMN_MAP = 'id:user_id,name:full_name';
SELECT id, name FROM Users;  -- Could generate INSERT INTO target(user_id, full_name)
```

However, this adds significant complexity and the current `SELECT *` approach covers most use cases.

### 4. Multi-Statement Support
Currently, each statement is processed independently. Future versions could detect related statements:
```sql
-- Future: detect table from preceding statement
CREATE TABLE Target LIKE Source;
SELECT * FROM Source;  -- Could infer Target as destination
```

### 5. Integration with DUMP Statements
The auto-detection logic could potentially be shared with DUMP statements to provide consistent table name handling across features.

These enhancements are not currently planned but documented here for future reference if use cases emerge that would benefit from them.
